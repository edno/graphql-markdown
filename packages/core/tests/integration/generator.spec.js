const { vol } = require("memfs");
jest.mock("fs");

jest.mock("@graphql-markdown/printer-legacy", () => {
  return {
    printType: jest.fn(),
    init: jest.fn(),
  };
});
const Printer = require("@graphql-markdown/printer-legacy");

jest.mock("@graphql-markdown/diff");
const diff = require("@graphql-markdown/diff");

const { generateDocFromSchema } = require("../../src/generator");

describe("renderer", () => {
  beforeEach(() => {
    jest.spyOn(Printer, "printType").mockImplementation((value) => value);

    vol.fromJSON({
      "/output": {},
      "/temp_test": {},
      "/assets/generated.md": "Dummy homepage for tweet.graphql",
    });
  });

  afterEach(() => {
    vol.reset();
    jest.restoreAllMocks();
  });

  describe("generateDocFromSchema()", () => {
    test("generates Markdown document structure from GraphQL schema", async () => {
      expect.assertions(1);

      const config = {
        baseURL: "graphql",
        schemaLocation: "tests/__data__/tweet.graphql",
        outputDir: "/output",
        linkRoot: "docs",
        homepageLocation: "/assets/generated.md",
        diffMethod: "NONE",
        tmpDir: "/temp",
        loaders: { GraphQLFileLoader: "@graphql-tools/graphql-file-loader" },
        printer: "@graphql-markdown/printer-legacy",
      };

      await generateDocFromSchema(config);

      expect(vol.toJSON(config.outputDir, undefined, true)).toMatchSnapshot();
    });

    test('outputs "no schema changed" message when called twice', async () => {
      expect.assertions(1);

      const logSpy = jest.spyOn(console, "info");

      const config = {
        baseURL: "graphql",
        schemaLocation: "tests/__data__/tweet.graphql",
        outputDir: "/output",
        linkRoot: "docs",
        homepageLocation: "/assets/generated.md",
        diffMethod: "SCHEMA-HASH",
        tmpDir: "/temp",
        loaders: { GraphQLFileLoader: "@graphql-tools/graphql-file-loader" },
        printer: "@graphql-markdown/printer-legacy",
      };

      jest.spyOn(diff, "checkSchemaChanges").mockReturnValue(false);
      await generateDocFromSchema(config);

      expect(logSpy).toHaveBeenCalledWith(
        `No changes detected in schema "${config.schemaLocation}".`,
      );
    });

    test("Markdown document structure from GraphQL schema is correct when using grouping", async () => {
      expect.assertions(2);

      const config = {
        baseURL: "graphql",
        schemaLocation: "tests/__data__/schema_with_grouping.graphql",
        outputDir: "/output",
        linkRoot: "docs",
        homepageLocation: "/assets/generated.md",
        diffMethod: "NONE",
        tmpDir: "/temp",
        loaders: { GraphQLFileLoader: "@graphql-tools/graphql-file-loader" },
        groupByDirective: {
          directive: "doc",
          field: "category",
          fallback: "misc",
        },
        printer: "@graphql-markdown/printer-legacy",
      };

      await generateDocFromSchema(config);

      expect(vol.toJSON(config.outputDir, undefined, true)).toMatchSnapshot();
      expect(vol.toJSON(config.tmpDir, undefined, true)).toMatchSnapshot();
    });
  });
});
