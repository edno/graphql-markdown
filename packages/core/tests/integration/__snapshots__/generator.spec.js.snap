// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderer generateDocFromSchema() Markdown document structure from GraphQL schema is correct when using grouping 1`] = `
{
  "course/_category_.yml": "label: Course
position: 1
link: null
",
  "course/mutations/_category_.yml": "label: Mutations
position: 1
link: null
",
  "course/mutations/add-course.mdx": "add-course",
  "course/mutations/drop-course.mdx": "drop-course",
  "course/queries/_category_.yml": "label: Queries
position: 1
link: null
",
  "course/queries/all-courses.mdx": "all-courses",
  "course/queries/course.mdx": "course",
  "course/queries/math-courses.mdx": "math-courses",
  "course/queries/science-courses.mdx": "science-courses",
  "generated.md": "Dummy homepage for tweet.graphql",
  "grade/_category_.yml": "label: Grade
position: 1
link: null
",
  "grade/mutations/_category_.yml": "label: Mutations
position: 1
link: null
",
  "grade/mutations/update-gpa.mdx": "update-gpa",
  "grade/mutations/update-grade.mdx": "update-grade",
  "grade/queries/_category_.yml": "label: Queries
position: 1
link: null
",
  "grade/queries/gpa.mdx": "gpa",
  "grade/queries/un-weighted-gpa.mdx": "un-weighted-gpa",
  "grade/queries/weighted-gpa.mdx": "weighted-gpa",
  "misc/_category_.yml": "label: Misc
position: 1
link: null
",
  "misc/directives/_category_.yml": "label: Directives
position: 1
link: null
",
  "misc/directives/deprecated.mdx": "deprecated",
  "misc/directives/doc.mdx": "doc",
  "misc/directives/include.mdx": "include",
  "misc/directives/no-doc.mdx": "no-doc",
  "misc/directives/skip.mdx": "skip",
  "misc/directives/specified-by.mdx": "specified-by",
  "misc/enums/_category_.yml": "label: Enums
position: 1
link: null
",
  "misc/enums/roles.mdx": "roles",
  "misc/interfaces/_category_.yml": "label: Interfaces
position: 1
link: null
",
  "misc/interfaces/record.mdx": "record",
  "misc/objects/_category_.yml": "label: Objects
position: 1
link: null
",
  "misc/objects/course.mdx": "course",
  "misc/objects/department-information.mdx": "department-information",
  "misc/objects/semester.mdx": "semester",
  "misc/objects/unused-type.mdx": "unused-type",
  "misc/queries/_category_.yml": "label: Queries
position: 1
link: null
",
  "misc/queries/search-role.mdx": "search-role",
  "misc/scalars/_category_.yml": "label: Scalars
position: 1
link: null
",
  "misc/scalars/boolean.mdx": "boolean",
  "misc/scalars/date.mdx": "date",
  "misc/scalars/email-address.mdx": "email-address",
  "misc/scalars/id.mdx": "id",
  "misc/scalars/int.mdx": "int",
  "misc/scalars/phone-number.mdx": "phone-number",
  "misc/scalars/string.mdx": "string",
  "misc/scalars/time.mdx": "time",
  "misc/scalars/url.mdx": "url",
  "sidebar-schema.js": "
/** @type {import('@docusaurus/plugin-content-docs').SidebarsConfig} */

module.exports = {
  "schemaSidebar": [
    {
      "type": "autogenerated",
      "dirName": "graphql"
    }
  ]
};
",
}
`;

exports[`renderer generateDocFromSchema() Markdown document structure from GraphQL schema is correct when using grouping 2`] = `{}`;

exports[`renderer generateDocFromSchema() generates Markdown document structure from GraphQL schema 1`] = `
{
  "directives/_category_.yml": "label: Directives
position: 1
link: null
",
  "directives/deprecated.mdx": "deprecated",
  "directives/include.mdx": "include",
  "directives/skip.mdx": "skip",
  "directives/specified-by.mdx": "specified-by",
  "enums/_category_.yml": "label: Enums
position: 1
link: null
",
  "enums/media-type.mdx": "media-type",
  "generated.md": "Dummy homepage for tweet.graphql",
  "inputs/_category_.yml": "label: Inputs
position: 1
link: null
",
  "inputs/tweet-content.mdx": "tweet-content",
  "interfaces/_category_.yml": "label: Interfaces
position: 1
link: null
",
  "interfaces/node.mdx": "node",
  "mutations/_category_.yml": "label: Mutations
position: 1
link: null
",
  "mutations/create-tweet.mdx": "create-tweet",
  "mutations/delete-tweet.mdx": "delete-tweet",
  "mutations/mark-tweet-read.mdx": "mark-tweet-read",
  "objects/_category_.yml": "label: Objects
position: 1
link: null
",
  "objects/media.mdx": "media",
  "objects/meta.mdx": "meta",
  "objects/notification.mdx": "notification",
  "objects/stat.mdx": "stat",
  "objects/tweet.mdx": "tweet",
  "objects/user.mdx": "user",
  "queries/_category_.yml": "label: Queries
position: 1
link: null
",
  "queries/notifications-meta.mdx": "notifications-meta",
  "queries/tweet.mdx": "tweet",
  "queries/tweets-meta.mdx": "tweets-meta",
  "queries/tweets.mdx": "tweets",
  "queries/user.mdx": "user",
  "queries/users.mdx": "users",
  "scalars/_category_.yml": "label: Scalars
position: 1
link: null
",
  "scalars/boolean.mdx": "boolean",
  "scalars/date.mdx": "date",
  "scalars/id.mdx": "id",
  "scalars/int.mdx": "int",
  "scalars/sri.mdx": "sri",
  "scalars/string.mdx": "string",
  "scalars/url.mdx": "url",
  "sidebar-schema.js": "
/** @type {import('@docusaurus/plugin-content-docs').SidebarsConfig} */

module.exports = {
  "schemaSidebar": [
    {
      "type": "autogenerated",
      "dirName": "graphql"
    }
  ]
};
",
  "subscriptions/_category_.yml": "label: Subscriptions
position: 1
link: null
",
  "subscriptions/notifications.mdx": "notifications",
}
`;
